### ISSUE TITLE
LoginAsync method parameters changed break example code

### ISSUE THREAD

**Issue by JPhilC** [adrianhall/develop-mobile-apps-with-csharp-and-azure#74]
Hello Adrian,

I think that your sample code may now be out of date with the latest builds of the mobile service client and possibly the client flow for authentication and authorisation with MicrosoftAccount.

### Where is the problem?

https://adrianhall.github.io/develop-mobile-apps-with-csharp-and-azure/chapter2/social/#client-flow-for-social-providers

### What is the problem?
Taking the DroidLoginProvider code as an example, 

Calls to ...

await client.LoginAsync(context, "Facebook");

are no longer valid because there is no version of LoginAsync that takes those parameter types. The method now expects parameter types MobileServiceAuthenticationProvider  and a JToken object.

### Did you get an exception?

Just IDE error Argument 1: cannot convert from 'Android.Content.Context' to 'Microsoft.WindowsAzure.MobileServices.MobileServiceAuthenticationProvider'

### What platform are you compiling for?
.NET 4.7 (A desktop client)
Android 7.1 Nougat
iOS 10.0

I am also using a .NET Standard 2.0 class library to share models, abstractions and other code between the mobile and desktop clients.

### Package Versions
  <package id="CommonServiceLocator" version="1.3" targetFramework="monoandroid71" />
  <package id="Microsoft.Azure.Mobile.Client" version="4.0.1" targetFramework="monoandroid71" />
  <package id="Microsoft.Bcl" version="1.1.10" targetFramework="monoandroid71" />
  <package id="Microsoft.Bcl.Build" version="1.0.21" targetFramework="monoandroid71" />
  <package id="Microsoft.CSharp" version="4.4.0" targetFramework="monoandroid71" />
  <package id="Microsoft.Net.Http" version="2.2.29" targetFramework="monoandroid71" />
  <package id="Microsoft.NETCore.Platforms" version="2.0.0" targetFramework="monoandroid71" />
  <package id="Microsoft.Win32.Primitives" version="4.3.0" targetFramework="monoandroid71" />
  <package id="MvvmLightLibs" version="5.3.0.0" targetFramework="monoandroid71" />
  <package id="NETStandard.Library" version="2.0.0" targetFramework="monoandroid71" />
  <package id="Newtonsoft.Json" version="10.0.3" targetFramework="monoandroid71" />
  <package id="PCLCrypto" version="2.0.147" targetFramework="monoandroid71" />
  <package id="PInvoke.BCrypt" version="0.5.97" targetFramework="monoandroid71" />
  <package id="PInvoke.Kernel32" version="0.5.97" targetFramework="monoandroid71" />
  <package id="PInvoke.NCrypt" version="0.5.97" targetFramework="monoandroid71" />
  <package id="PInvoke.Windows.Core" version="0.5.97" targetFramework="monoandroid71" />
  <package id="System.AppContext" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Collections" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Collections.Concurrent" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.ComponentModel.TypeConverter" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Console" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Diagnostics.Debug" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Diagnostics.Tools" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Diagnostics.Tracing" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Globalization" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Globalization.Calendars" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.IO" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.IO.Compression" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.IO.Compression.ZipFile" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.IO.FileSystem" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.IO.FileSystem.Primitives" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Linq" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Linq.Expressions" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Net.Http" version="4.3.2" targetFramework="monoandroid71" />
  <package id="System.Net.Primitives" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Net.Sockets" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.ObjectModel" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Reflection" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Reflection.Extensions" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Reflection.Primitives" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Resources.ResourceManager" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Runtime" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Runtime.Extensions" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Runtime.Handles" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Runtime.InteropServices" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Runtime.InteropServices.RuntimeInformation" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Runtime.Numerics" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Runtime.Serialization.Formatters" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Runtime.Serialization.Primitives" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Security.Cryptography.Algorithms" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Security.Cryptography.Encoding" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Security.Cryptography.Primitives" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Security.Cryptography.X509Certificates" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Text.Encoding" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Text.Encoding.Extensions" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Text.RegularExpressions" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Threading" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Threading.Tasks" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Threading.Timer" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Xml.ReaderWriter" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Xml.XDocument" version="4.3.0" targetFramework="monoandroid71" />
  <package id="System.Xml.XmlDocument" version="4.3.0" targetFramework="monoandroid71" />
  <package id="Validation" version="2.4.15" targetFramework="monoandroid71" />
  <package id="Xamarin.Android.Support.Animated.Vector.Drawable" version="25.4.0.1" targetFramework="monoandroid71" />
  <package id="Xamarin.Android.Support.Annotations" version="25.4.0.1" targetFramework="monoandroid71" />
  <package id="Xamarin.Android.Support.Compat" version="25.4.0.1" targetFramework="monoandroid71" />
  <package id="Xamarin.Android.Support.Core.UI" version="25.4.0.1" targetFramework="monoandroid71" />
  <package id="Xamarin.Android.Support.Core.Utils" version="25.4.0.1" targetFramework="monoandroid71" />
  <package id="Xamarin.Android.Support.CustomTabs" version="25.4.0.1" targetFramework="monoandroid71" />
  <package id="Xamarin.Android.Support.Design" version="25.4.0.1" targetFramework="monoandroid71" />
  <package id="Xamarin.Android.Support.Fragment" version="25.4.0.1" targetFramework="monoandroid71" />
  <package id="Xamarin.Android.Support.Media.Compat" version="25.4.0.1" targetFramework="monoandroid71" />
  <package id="Xamarin.Android.Support.Transition" version="25.4.0.1" targetFramework="monoandroid71" />
  <package id="Xamarin.Android.Support.v4" version="25.4.0.1" targetFramework="monoandroid71" />
  <package id="Xamarin.Android.Support.v7.AppCompat" version="25.4.0.1" targetFramework="monoandroid71" />
  <package id="Xamarin.Android.Support.v7.CardView" version="25.4.0.1" targetFramework="monoandroid71" />
  <package id="Xamarin.Android.Support.v7.MediaRouter" version="25.4.0.1" targetFramework="monoandroid71" />
  <package id="Xamarin.Android.Support.v7.Palette" version="25.4.0.1" targetFramework="monoandroid71" />
  <package id="Xamarin.Android.Support.v7.RecyclerView" version="25.4.0.1" targetFramework="monoandroid71" />
  <package id="Xamarin.Android.Support.Vector.Drawable" version="25.4.0.1" targetFramework="monoandroid71" />
  <package id="Xamarin.Forms" version="2.3.4.267" targetFramework="monoandroid71" />

### What did you do?

Spent many hours trawling the internet trying to workout how to get this to work. I am guessing that the new method expects you to get an authentication token before calling LoginAsync but I have been unable to find a clear explanation of how to do this anywhere. I was hoping to be able to authenticate and authorise using MicrosoftAccount.

Thanks and I really appreciate book.

Phil Crompton

**Comment by adrianhall** [comment]
A quick look at the code suggests you are right.  Unfortunately, nothing I can do about it.  I am not currently working on this book, and SDKs change all the time (which is why I didn't publish this book in paper form).  I'd take this bug up with the Microsoft team.

**Comment by JPhilC** [comment]
Thanks for the quick reply. I have posted an issue for the Microsoft team as you suggest.

**Comment by Mephisztoe** [comment]
It's 2019 now and right yesterday, I stumbled across this problem as well. Also, this book seems to be still THE best source for explaining a best practice of working with Azure App Services. So I am trying to refactor my work.

It's really weird... Because I was used to this practice, I implemented repositories and services on client side just to notice that when incorporating the natural behaviour of the Azure Mobile Apps Client SDK, I practically do not need any of this...

I hope that a new version of this book is not going to change too much of this, because getting everything to work now is already a challenge...
